File: denoise_saltpepper.m
clc
clear all
close all

% Đọc ảnh, chuyển sang ảnh xám, và crop ảnh nếu kích thước lớn hơn 255
I = imread('dataset\4.1.03.tiff');
I = rgb2gray(I);
[rows, cols] = size(I);
size = [rows, cols];
if (rows > 256), I = imresize(I, [256 Nan]), end;
if (cols > 256), I = imresize(I, [Nan 256]), end;

figure
tiledlayout(3,3)

% Hiển thị ảnh gốc
nexttile, imshow(I), title('Anh goc')

% Thêm nhiễu hạt với mật độ 0.05 và hiển thị ảnh
noise = rand(size);
J = zeros(size);
J = uint8(J) + I .* uint8(noise > 0.2) + 255 .* uint8(noise < 0.1);
nexttile, imshow(J), title('Nhieu muoi tieu')

% Lọc trung vị và hiển thị ảnh
Kcross5 = medFilterCross5(J);
nexttile, imshow(Kcross5), title('Loc trung vi chu thap 5 diem')

% Lọc trung vị bằng hàm có sẵn medfill2
K2x2 = medfilt2(J, [2 2]);
nexttile, imshow(K2x2), title('Loc trung vi kich thuoc 2x2')
K3x3 = medfilt2(J, [3 3]);
nexttile, imshow(K3x3), title('Loc trung vi kich thuoc 3x3')
K4x4 = medfilt2(J, [4 4]);
nexttile, imshow(K4x4), title('Loc trung vi kich thuoc 4x4')

% Hàm lọc trung vị dạng 5 điểm chữ thập (5-point cross-shaped)
function out = medFilterCross5(in)
    [rows, cols] = size(in);
    out = zeros([rows, cols]);
    for i = 2:rows-1
        for j = 2:cols-1
            x = [in(i,j), in(i-1,j), in(i+1,j), ...
                in(i,j-1), in(i,j+1)];
            out(i,j) = median(x);
        end
    end
    out = uint8(out);
end

% Sai số trung bình bình phương
function out = myMSE(im1, im2)
    [rows, cols] = size(im1);
    error = im1 - im2;
    out = sum(error.^2, 'all') / (rows * cols);
end

File: denoise_saltpepper.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from scipy.ndimage import median_filter

# Đọc ảnh và chuyển sang ảnh xám
I = cv2.imread('dataset/4.1.03.tiff', cv2.IMREAD_GRAYSCALE)
nRows, nCols = I.shape
size = (nRows, nCols)

fig, axs = plt.subplots(2, 3)

# Vẽ ảnh gốc
axs[0][0].imshow(I, cmap='gray', vmin=0, vmax=255)
axs[0][0].set_title('Ảnh gốc')

# Tạo nhiễu muối tiêu với mật độ 0.05 và hiển thị ảnh
noise = np.random.random(size)
J = np.zeros(size, dtype=int)
J = J + I * (noise > 0.2) + 255 * (noise < 0.1)
axs[0][1].imshow(J, cmap='gray', vmin=0, vmax=255)
axs[0][1].set_title('Nhiễu muối tiêu')

# Hàm lọc trung vị dạng 5 điểm chữ thập (5-point cross-shaped)
def medFilterCross5(X):
  nRows, nCols = X.shape
  Y = np.zeros((nRows,nCols))
  for i in range(1,nRows-1):
    for j in range(1,nCols-1):
      x = (X[i][j], X[i-1][j], X[i+1][j], X[i][j-1], X[i][j+1])
      x = sorted(x)
      # Lấy trung vị bằng cách sắp xếp theo thứ tự lớn dần
      # và chọn phần tử thứ 3 trong 5 phần tử
      Y[i][j] = x[2]
  return Y

# Lọc trung vị 5 điểm chữ thập và hiển thị ảnh
Kcross5 = medFilterCross5(J)
axs[0][2].imshow(Kcross5, cmap='gray', vmin=0, vmax=255)
axs[0][2].set_title('Lọc trung vị \n5 điểm chữ thập')

# Lọc trung vị bằng hàm có sẵn scipy.ndimage.median_filter
K2x2 = median_filter(J, size=(2,2))
axs[1][0].imshow(K2x2, cmap='gray', vmin=0, vmax=255)
axs[1][0].set_title('Lọc trung vị \nkích thước 2x2')

K3x3 = median_filter(J, size=(3,3))
axs[1][1].imshow(K3x3, cmap='gray', vmin=0, vmax=255)
axs[1][1].set_title('Lọc trung vị \nkích thước 3x3')

K4x4 = median_filter(J, size=(4,4))
axs[1][2].imshow(K4x4, cmap='gray', vmin=0, vmax=455)
axs[1][2].set_title('Lọc trung vị \nkích thước 4x4')

for rowAxs in axs:
  for ax in rowAxs:
    ax.set_xticks([])
    ax.set_yticks([])

fig.tight_layout(h_pad=2)
fig.show()
fig.savefig("figures/MedianFiltering.png", bbox_inches='tight')
